### PR TITLE
Feat/training loop

### PR DESCRIPTION
**Pull request by EtienneChollet**
# Features
## Training Loggers (Commit 1180817)
- I put a logger class (`Logger`) into `cassetta/training/loggers.py` to evaluate various aspects of model training in real time. These have helped me tremendously in debugging my models and training loops.
- These logging features rely on tensor board, so tensor board would need to be included in the dependencies of this project (I'm not sure if they already are)
- Some `Logger` features include:
   - **Graph logging** (`Logger.log_model_graph()`)
      - This logs the graph of the backbone by passing a sample of data through it, to ensure data is flowing properly/the layers are connected properly.
    - **Model weight and gradient logging** (`Logger.log_parameter_histograms()`)
        - Log the frequency distributions of the weight and gradient magnitudes for each learnable parameter of the architecture, grouped by named module (helps debugging vanishing/exploding gradients, and if the model is not training —> no change in model weights)
    - Log training/validation metrics (`Logger.log_metrics()`)

## Training Loop (Commit 1d3a1d7)
- I decided to make the thing that handles all the training a class based object (`Trainer`), in a way similar to pytorch lightning.
- I have not yet added checkpointing functionalities - I plan to do this in the near future, once I have a better idea of how the `LoadableMixin` class saves stuff. Luckily, you made this class very intuitive, so it shouldn't take much time. I would like to get the directory structure sorted out first so I know where to save them (refer to the 'Ideas going forward' section).
- I also want to add a LR logger to this for when (or, if) you choose to make a variable LR schedule.

# Ideas going forward
- I'm not sure how you want to handle the structure for model saving and loading.
    - If you liked the way it was done in oct_vesselseg:
        - **Parent directories for different models/ideas/projects**.
            - `cassetta/models` : Generic parent directory for model saving
            - `cassetta/oct_seg_models` : Parent directory for some kind of segmentation model for oct.
            - `cassetta/oct_seg_with_nadam` : Parent directory for seg models for oct (perhaps for optimizing model configuration) using the nadam optimizer.
        - Under each project/configuration-specific directory, use versioned subdirectories to track different training runs or variations of the model:
            - `cassetta/oct_seg_with_nadam/version_1` : First training run of the nadam configuration
            - `cassetta/oct_seg_with_nadam/version_2` : Second training run of the nadam configuration
            - etc…

### CODE REVIEW

**Comment by balbasty** [comment]
Thanks Etienne!

There's quite a few builtin assumptions in the trainer, so we may want to use a more specific name to keep generic names like `Trainer` for future, potentially more general, classes. For example, your trainer expects the dataloader to yield two tensors (and they must be tensors since you call `to` on them), one passed to the model and one passed to the loss. It is quite restrictive: what if my model takes two tensors? (e.g., for registration), what if there is nothing to give to the loss? (unsupervised). Maybe we should call it `SimpleSupervisedTrainer` for now?

Also, I would maybe make logging optional (even if set to `True` by default).

In terms of model saving, the point of `LoadableMixIn` was to make it possible to save the parameters that were used to instantiate the model, not just the model weights, so that if we have a saved model we can directly instantiate and load it, without having to kmow all the parameters that were used initially.

We probably want two different ways of saving: 
- one during training that saves a lot of stuff (model weights and parameters, optimizer state and parameters, current state of training [epoch, step, etc] and training parameters), so that training can be stopped and restarted at will.
- one when we're happy that only saves the final/best model weights (with maybe an option for weight averaging) and parameters, which would be uploaded to (e.g.) figshare and used during deployment.

I know that's not what you're suggesting, but just in case I am not sure I would save model weights to gihub, but rather to figshare/zenodo. We could then have a utility to automatically download weights if they are not yet available locally. And we can also have an option in setup.py to download models during installation if the user requires it (for example, if they're not sure to have an internet connection at runtime). 

For the local directory structure during training, I am not entirely sure. I would do something like
```
{experiment_name}
|-- config.yml
|-- run_{d}
     |-- epoch-{d}.cpt
```
?

Let's talk more tomorrow!

**Comment by EtienneChollet** [comment]
Hey Yael, 

I made some of the changes we talked about:
1. Renaming `trainer.py` to `simple_trainer.py` (maybe I should remove that underscore)
2. Renaming `Trainer()` to `SimpleSupervisedTrainer()`
3. Added the option to log or not
4. Implemented some of the mixins from learn2synth like the `StateMixin`, `TrainerInfo()`, and `Counter()`

Let me know what you think, or if I missed anything!

File: cassetta/training/loggers.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
+__all__ = ['Logger']
+from torch import nn
+from torch.utils.tensorboard import SummaryWriter
+from torch.utils.data import DataLoader
+
+
+class Logger(object):
```
Comment (balbasty) [inline cassetta/training/loggers.py:7]: Is this a generic logger, or a tensorboard logger? If the latter I'd name it something like `TBLogger`

File: cassetta/training/loggers.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+        Parameters
+        ----------
+        epoch : int
+            Current epoch associated with state of parameters.
+        """
+        # Iterate through all named parameter groups
+        for name, param in self.model.named_parameters():
+            self.writer.add_histogram(name, param, epoch)
+            # TODO: Should we just log all?
```
Comment (balbasty) [inline cassetta/training/loggers.py:67]: What is this question about?

File: cassetta/training/simple_trainer.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        """
+        Train the model over num_epochs epochs.
+
+        Parameters
+        ----------
+        num_epochs : int
+            Number of epochs to train the model.
+        """
+        # Refreshing output directories (making them if they don't exist)
+        # TODO: Add something to refresh model saving directory.
```
Comment (balbasty) [inline cassetta/training/simple_trainer.py:113]: When would you refresh the directory? What kind of change do you expect?

File: cassetta/training/simple_trainer.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+        for epoch in range(self.trainer.nb_epochs):
+            train_loss = self._train_one_epoch()
+            val_loss = self._validate()
+            print(f"Epoch [{epoch+1}/{self.trainer.nb_epochs}], "
+                  f"Train Loss: {train_loss:.4f}, "
+                  f"Validation Loss: {val_loss:.4f}")
+            if self.trainer.enable_logging:
+                if self.counter.epoch % self.trainer.save_every == 0:
+                    # Log parameter and gradient frequency distributions
+                    self.logger.log_parameter_histograms(self.counter.epoch)
```
Comment (balbasty) [inline cassetta/training/simple_trainer.py:127]: Can we have an option to activate/deactivate the different types of logging? I'd assume it has a slight impact on training time?

File: cassetta/training/simple_trainer.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+            # Backward pass
+            loss.backward()  # Backprop
+            self.optimizer.step()  # Step the optimizer
+
+            epoch_train_loss += loss.item()  # Add loss to running train loss
+            # Log instantaneous training metrics every 10 steps
+            epoch_train_loss /= len(self.train_loader)
+            if self.trainer.enable_logging:
+                if i % 10 == 0:
```
Comment (balbasty) [inline cassetta/training/simple_trainer.py:166]: How was this `10` chosen? This should maybe be an option? 

Also, do you not print anything during training? Is it all through the logger?

File: cassetta/training/simple_trainer.py [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+            x0, zt = x0.to(self.device)[0], zt.to(self.device)[0]
+            self.optimizer.zero_grad()  # Zero parameter gradients
+
+            # Forward pass
+            outputs = self.model(zt)  # Forward pass through model
+            loss = self.criterion(x0, outputs)  # Calculate loss
+
+            # Backward pass
+            loss.backward()  # Backprop
+            self.optimizer.step()  # Step the optimizer
```
Comment (balbasty) [inline cassetta/training/simple_trainer.py:160]: I'd think that everything after this line should be wrapped in a `with torch.no_grad():` block.

Or maybe better, do `loss = loss.item()` here, and it will cut the computational graph.

Every-time you do `loss.item()` your sending data from the GPU to the CPU, which is slow, so better not to repeat it multiple times (currently you call `item()` twice)

File: cassetta/training/simple_trainer.py [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+
+    def _validate(self):
+        """
+        Evaluate model on validation set.
+
+        Returns
+        -------
+        val_loss : float
+            Average loss for validation loop.
+        """
```
Comment (balbasty) [inline cassetta/training/simple_trainer.py:193]: Instead of having big indented blocks, I like doing thinks like
```python
def _validate(self):
    # recursive context
    if torch.is_grad_enabled():
        with torch.nograd():
            return self._validate()
    # actual function code
    ...
```

**Review by balbasty — CHANGES_REQUESTED** [review]
It's geting there :)

I'd like to see it in action, can you add this dummy training script that we talked about?
It could be in `cassetts/tests/test_simple_trainer.py` and inside have the function

```python
def test_dummyrun():
    # some code that sets up a model, a trainer, and runs it for like 2 epoches
    ....
```
